// ng-zorro-antd/esm2015/result/nz-result-cells.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Alibaba.com All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { Directive } from '@angular/core';
export class NzResultTitleDirective {
}
NzResultTitleDirective.decorators = [
    { type: Directive, args: [{
                selector: 'div[nz-result-title]',
                exportAs: 'nzResultTitle',
                host: {
                    class: 'ant-result-title'
                }
            },] }
];
export class NzResultSubtitleDirective {
}
NzResultSubtitleDirective.decorators = [
    { type: Directive, args: [{
                selector: 'div[nz-result-subtitle]',
                exportAs: 'nzResultSubtitle',
                host: {
                    class: 'ant-result-subtitle'
                }
            },] }
];
export class NzResultIconDirective {
}
NzResultIconDirective.decorators = [
    { type: Directive, args: [{
                selector: 'i[nz-result-icon], div[nz-result-icon]',
                exportAs: 'nzResultIcon'
            },] }
];
export class NzResultContentDirective {
}
NzResultContentDirective.decorators = [
    { type: Directive, args: [{
                selector: 'div[nz-result-content]',
                exportAs: 'nzResultContent',
                host: {
                    class: 'ant-result-content'
                }
            },] }
];
export class NzResultExtraDirective {
}
NzResultExtraDirective.decorators = [
    { type: Directive, args: [{
                selector: 'div[nz-result-extra]',
                exportAs: 'nzResultExtra',
                host: {
                    class: 'ant-result-extra'
                }
            },] }
];
